README.md
Stochasm
========

A JavaScript component to create functions that generate random values.


Why?
----

It can be very useful to generate random numbers that are governed by properties of different types of distributions. Such distributions are useful for modeling numerical behavior and response of systems.



Fork
----

This module was forked from https://github.com/heydenberk/stochasm that was created by [Eric Heydenberk](http://twitter.com/heydenberk). Why did I fork it?

- Unmaintained
- Written in CoffeeScript
- Not a UMD
- No tests
- No support to modify the random number generation

Rather than pester Eric about changing any of the above, a fork seemed more reasonable.



Name
----

`stochasm` is a portmanteau of stochastic and chasm.



Install
-------

### Node.js/Browserify

    npm install --save stochasm


### Component

    component install jprichardson/stochasm


### Bower

    bower install stochasm


### Script

```html
<script src="/path/to/stochasm.js"></script>
```


Usage
-----

To create a `stochasm` object, simply invoke the function and pass it an `options` object with a `kind` property. If not provided, kind is 'float'.

Valid kinds include `float`, `integer`, `set`.


### Floating-point Decimals

It's very easy generate a float between 0 and 1.

````js
var stochasm = require('stochasm')

var generator = stochasm()
generator.next(); // 0.9854211050551385
generator.next(); // 0.8784450970124453
generator.next(); // 0.1592887439765036
````

This is not very exciting because it simply wraps the built-in `Math.random` method.



### Floats from an Interval

Specifying a min and a max allows us to create random numbers in the interval (min, max), not inclusive.

````js
var radianGenerator = stochasm({min: 0, max: Math.PI * 2})
radianGenerator.next(); // 3.7084574239999655
radianGenerator.next(); // 1.021138034566463
radianGenerator.next(); // 4.012664264853087
````



### Floats from a Normal Distribution

We can also generate random floats from a normal distribution. Min and max are optional, and when provided will result in truncation of all results outside of [min, max].

````js
var testScores = stochasm({mean: 75, stdev: 14, min: 0, max: 100})
testScores.next(); // 59.437160028200125
testScores.next(); // 80.18612670399554
testScores.next(); // 75.81242027226946
````



### Integers

For integers, the interval [min, max] is inclusive.

````js
var die = stochasm({kind: "integer", min: 1, max: 6})
die.next(); // 6
die.next(); // 1
die.next(); // 2
````

if `next()` feels out of place for your use case, just rename the method:

```js
die.roll = die.next
die.roll() //4
```

## Multiple Results
If the `next` method (or a method aliased to it) is passed an integer `n`, it will return an n-length array of results. Using the die instance from the previous example:

````js
die.roll(1); // [5]
die.roll(2); // [5, 3]
die.roll(5); // [6, 3, 6, 6, 5]
````


### From Sets

We can generate random values from arbitary sets.

````js
var dayGenerator = new stochasm({
	kind: "set",
	values: ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"]
});
dayGenerator.next(); // friday
dayGenerator.next(); // monday 
dayGenerator.next(); // monday
````
	


### From Sets with Weights

What if we favor the weekend? Well, we can pass `weights`, an array of the same length as `values` consisting of probabilities out of 1 that correspond to `values`.

````js
var biasedDayGenerator = new stochasm({
	kind: "set",
	values: ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"],
	weights: [0.1, 0.1, 0.1, 0.1, 0.1, 0.25, 0.25]
});
biasedDayGenerator.next(); // thursday
biasedDayGenerator.next(); // sunday 
biasedDayGenerator.next(); // saturday
````



### From Sets Without Replacement

Note: This functionality may be removed.

Passing a `replacement` property with a falsy value will result in each random
value generation to be removed from the set.

````js
var chores = new stochasm({
	kind: "set",
	values: ["floors", "windows", "dishes"],
	replacement: false
});
var myChore = chores.next(); // "windows"
var yourChore = chores.next(); // "floors"
var hisChore = chores.next(); // "dishes"
var noOnesChore = chores.next(); // undefined
````



### Mutators


The constructor accepts an optional final argument which is passed the output
of the random value generator. Its return value becomes the return value of
next or its alias. To generate random boolean values, we can do:

````js
var booleanGenerator = new stochasm({
	kind: "integer",
	min: 0,
	max: 1
}, Boolean);

booleanGenerator.next(); // false
booleanGenerator.next(); // true
booleanGenerator.next(); // true
````

We can map the previously mentioned `radianGenerator` to the cosine of its values.

````js
var radianSineGenerator = new stochasm({
	min: 0,
	max: Math.PI * 2
}, Math.cos);
radianSineGenerator.next(); // -0.31173382958096524
radianSineGenerator.next(); // -0.6424354006937544
radianSineGenerator.next(); // 0.6475980728835664
````

Mutators remember their previous result and, at each generation, apply the results of a specified stochasm to create a new result.

 _(This is functionally equivalent to a Markov chain.)_

````js
var drunkardsWalk = new stochasm({
	kind: "integer",
	min: -1,
	max: 1
}, function(a, b) { return a + b; });

drunkardsWalk.value = 0; // Sets the initial value
drunkardsWalk.next(10); // [-1, -2, -2, -1, -1, -1, 0, 1, 1, 2]
drunkardsWalk.next(10); // [3, 3, 3, 2, 1, 0, -1, 0, 0, 0]
drunkardsWalk.next(10); // [0, 1, 0, -1, 0, 0, 1, 2, 1, 1]
````

Let's model a bank account's balance. How much money might you have after 10 years if you start with $1000, add $1000 every year, and get interest at a random rate between 1% and 5%?

````js
var addInterest = function(interestRate, principal) {
	return (principal + 1000) * interestRate;
};
var savingsAccountBalance = new stochasm({
	kind: "float",
	min: 1.01,
	max: 1.05
}, addInterest);

savingsAccountBalance.value = 1000; // Sets the initital value
savingsAccountBalance.next(10);
/*
[
	2096.2402432970703,
	3177.3792999428224,
	4339.349049328612,
	5441.863800747634,
	6507.916293297546,
	7669.519280743041,
	9011.783840249629,
	10225.82489660009,
	11630.122217972781,
	12782.667463879243
]
*/
````



### Multiple Generators

If the stochasm function is passed multiple configuration objects, `next` (or its alias) returns an array of each random generated value.

To generate a random point, we might do:

````js
var x = { kind: 'integer', min: 0, max: 480 };
var y = { kind: 'integer', min: 0, max: 360 };
var mutator = function(values) {
	return {
		x: values[0],
		y: values[1]
	};
};
var randomPoint = new stochasm(x, y, mutator);

randomPoint.next(); // { x: 79, y: 65 }
randomPoint.next(); // { x: 151, y: 283 }
randomPoint.next(); // { x: 5, y: 253 }
````


### Your Own Random Number Generator

Want to bring your own random number generator to the party? Whether you're working with Node.js and want to use `crypto.getRandomValues()` or in the browser and want to use `window.crypto.getRandomValues()`, you can. You could also return a constant to see how your system may respond to certain conditions in testing.

Note: At this time, it's assumed that any function that you set `rand` to will return a number in the range of `[0,1)`.


```js
var stochasm = require('stochasm')

var radianGenerator = stochasm({min: 0, max: Math.PI * 2})
radianGenerator.rand = function() { return 0.4 }
```


Credits
-------

As stated above, this code was forked from the Node.js module `stochator` (https://github.com/heydenberk/stochator) that was created by [Eric Heydenberk](http://twitter.com/heydenberk). Eric Heydenberk deserves much of the credit for coming up with such an awesome idea.



License
-------

It made sense to just keep the license the same. MPLv2.

(MPLv2 License)

Copyright 2013, Eric Heydenberk and JP Richardson




CHANGELOG.md
0.5.0 / 2013-10-11
------------------
- bring your own random number genator
- removed the capability of renaming the `next` function
- `bower`, `component`, `require.js` support

0.4.0 / 2013-10-10
------------------
- forked from `stochator`
- removed CoffeeScript
- remove kinds `color` and `a-z`

test/util.js
var me = module.exports

me.mean = function(vals) {
  return vals.reduce(function(acc, cur) { return acc + cur },0) / vals.length
}

me.stdev = function (vals) {
  var mean = me.mean(vals)
  var sumSq = vals.reduce(function(acc, cur) { return acc + Math.pow(cur - mean, 2)},0)
  return Math.sqrt(sumSq / (vals.length - 1))
}

test/stochasm.test.js
var should = require('should')
  , stochasm = require('../lib/stochasm')
  , u = require('./util')
  , _ = require('lodash')

var SAMPLE_SIZE = 10000//00

describe('+ stochasm', function() {
  describe('> when no parameters', function() {
    it('should create a function that generates values between 0.0 and 1.0', function() {
      var generator = stochasm()
      var vals = []
      for (var i = 0; i < SAMPLE_SIZE; ++i) {
        vals.push(generator.next())
      }

      u.mean(vals).should.be.approximately(0.5, 0.1)
      _(vals).min().should.be.approximately(0, 0.1)
      _(vals).max().should.be.approximately(1, 0.1)

      var newVals = generator.next(SAMPLE_SIZE)
      newVals.length.should.eql(SAMPLE_SIZE)
    })
  })

  describe('> when min/max interval parameters are passed', function() {
    it('should create a function that generates values between the min and max', function() {
      var generator = stochasm({min: -Math.PI, max: 2 * Math.PI})
      var vals = []
      for (var i = 0; i < SAMPLE_SIZE; ++i) {
        vals.push(generator.next())
      }

      u.mean(vals).should.be.approximately(Math.PI / 2, 0.1)
      _(vals).min().should.be.approximately(-Math.PI, 0.1)
      _(vals).max().should.be.approximately(2 * Math.PI, 0.1)
    })
  })

  describe('> when min, max, mean, and stddev parameters are passed', function() {
    it('should create a function that generates values governed by the normal distribuation', function() {
      var sd = 14
      var m = 75

      var generator = stochasm({mean: m, stdev: sd, min: 0, max: 100})
      var vals = []
      for (var i = 0; i < SAMPLE_SIZE; ++i) {
        vals.push(generator.next())
      }

      u.mean(vals).should.be.approximately(75, 1)
      u.stdev(vals).should.be.approximately(14, 1)
      _(vals).min().should.be.approximately(0, 2*sd)
      _(vals).max().should.be.approximately(100, 1)

    })
  })

  //will probably remove the ability to use "kinds" and to use custom named functions
  describe('> when the kind is an integer', function() {
    it('should create a function that generates bounded integers', function() {
      var generator = stochasm({kind: "integer", min: 1, max: 6})
      generator.roll = generator.next
      var vals = []
      for (var i = 0; i < SAMPLE_SIZE; ++i) {
        vals.push(generator.roll())
      }
 
      _.min(vals).should.equal(1) 
      _.max(vals).should.equal(6)

      //is it a fair die? hehehe
      var counts = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6:0}
      vals.forEach(function(n) {
        counts[n] += 1
      })

      Object.keys(counts).forEach(function(n) {
        var count = counts[n]
        count.should.be.approximately(SAMPLE_SIZE / 6, SAMPLE_SIZE*0.03)
      })
    })
  })

  describe('> when the kind is a set', function() {
    it('should create a function that generates set items', function() {
      var dayGenerator = stochasm({
        kind: "set",
        values: ["m", "t", "w", "r", "f", "sa", "su"]
      })
     
      var vals = []
      for (var i = 0; i < SAMPLE_SIZE; ++i) {
        vals.push(dayGenerator.next())
      }

      var counts = {'m': 0, 't': 0, 'w': 0, 'r': 0, 'f': 0, 'sa': 0, 'su': 0}
      vals.forEach(function(d) {
        counts[d] += 1
      })

      Object.keys(counts).forEach(function(d) {
        var count = counts[d]
        count.should.be.approximately(SAMPLE_SIZE / 7, SAMPLE_SIZE*0.03)
      })      
    })
  })

  describe('> when the kind is a set with weights', function() {
    it('should create a function that generates set items according to weight', function() {
      var dayGenerator = stochasm({
        kind: "set",
        values: ["m", "t", "w", "r", "f", "sa", "su"],
        weights: [0.1, 0.1, 0.1, 0.1, 0.1, 0.25, 0.25] //favor weekends
      })
      
      var vals = []
      for (var i = 0; i < SAMPLE_SIZE; ++i) {
        vals.push(dayGenerator.next())
      }

      var counts = {'m': 0, 't': 0, 'w': 0, 'r': 0, 'f': 0, 'sa': 0, 'su': 0}
      vals.forEach(function(d) {
        counts[d] += 1
      })

      Object.keys(counts).forEach(function(d) {
        if (d == 'sa' || d == 'su') return

        var count = counts[d]
        count.should.be.approximately(SAMPLE_SIZE*0.1, SAMPLE_SIZE*0.03)
      })

      counts['sa'].should.be.approximately(SAMPLE_SIZE*0.25, SAMPLE_SIZE*0.03)
      counts['su'].should.be.approximately(SAMPLE_SIZE*0.25, SAMPLE_SIZE*0.03)
    })
  })

  describe('> when a mutator function is specified', function() {
    it('should create a function that generates values according to the mutator', function() {
      function mutate (val) {
        return val >= 0 ? '+' : '-' //slight bias towards positive because ">= 0"
      }

      var pnGen = stochasm({kind: 'float', min: -1.0, max: 1.0}, mutate)
      var vals = pnGen.next(SAMPLE_SIZE)

      var counts = {'-': 0, '+': 0}
      vals.forEach(function(v) { counts[v] += 1 })

      Object.keys(counts).forEach(function(v) {
        var count = counts[v]
        count.should.be.approximately(SAMPLE_SIZE / 2, SAMPLE_SIZE*0.03)
      }) 
    })
  })

  describe('> when a mutator function is specified with an initial value', function() {
    it('should create a function that generates values according to the mutator and initial value', function() {
      var min = 1.01
      var max = 1.05
      var initVal = 1000
      var annualSavings = 250

      //models a bank account: how much money after 10 years if you add 250 a month, have a random interest rate from 1% to 5%
      function addInterest(interestRate, principal) {
        return (principal + annualSavings) * interestRate
      }

      var savingsAccount = stochasm({kind: 'float', min: min, max: max}, addInterest)
      savingsAccount.value = initVal
      
      var amounts = savingsAccount.next(10) //amounts for 10 years
      for (var i = 0; i < amounts.length; ++i) {
        var err = Math.pow(1 + (max - min)/2, 1+i)
        var x = (initVal + (i+1)*annualSavings) * err
        //console.log("real %d should be close to estimate %d", amounts[i], x)
        amounts[i].should.be.approximately(x, annualSavings*err*1.03)
      }
    })
  })

  describe('> when multiple generators are are input', function() {
    it('should create a function that generates multiple values governed by the configuration', function() {
      var x = { kind: 'integer', min: 0, max: 1400 };
      var y = { kind: 'integer', min: 0, max: 900 };
      var mutator = function(values) {
        return {
          x: values[0],
          y: values[1]
        }
      }

      var randomPoint = stochasm(x, y, mutator)
      var pts = randomPoint.next(SAMPLE_SIZE)

      _.min(pts, function(pt) { return pt.x }).x.should.be.approximately(0, 10)
      _.min(pts, function(pt) { return pt.y }).y.should.be.approximately(0, 10)
      _.max(pts, function(pt) { return pt.x }).x.should.be.approximately(1400, 10)
      _.max(pts, function(pt) { return pt.y }).y.should.be.approximately(900, 10)
    })
  })
})




lib/set.js
//am going to slowly delete this module

module.exports = Set

function Set(values) {
  this.values = values;
  this.length = this.values.length;
}

Set.prototype.copy = function() {
  return this.values.slice(0)
}

Set.prototype.enumerate = function(depth) {
  var len = this.values.length
  if (depth == null) depth = len
  var enumerationsLength = Math.pow(len, depth);
  var enumerations = [];
  
  for (var e = 0; e < enumerationsLength; ++e) {
    var digits = []
    for (var i = 0; i < depth; ++i) {
      var d = e % len
      e -= d;
      e /= len
      digits.push(this.values[d]); 
    }
    enumerations.push(new Set(digits));
  }

  return new Set(enumerations)
}

Set.prototype.reduce = function(iterator) {
  return this.values.reduce(iterator);
}

Set.prototype.sum = function() {
  if (this._sum) return _sum
  this._sum = this.reduce(function(a, b) { return a + b })
  return this._sum
}

Set.prototype.get = function(index, defaultVal) {
  if (this.values[index] == null) return defaultVal
  return this.values[index]
}

Set.prototype.each = function(iterator) {
  return this.values.map(iterator)
}

Set.prototype.map = function(iterator) {
  return new Set(this.each(iterator))
}

Set.prototype.pop = function(index) {
  if (index == null) index = this.length - 1
  
  value = this.values[index]
  this.values.splice(index, 1)
  this.length -= 1 //<--- this is not in the original code, but it should
  return value
}






lib/unused.js
randomColor = function() {
  var byte, mutator;
  byte = {
    kind: "integer",
    min: 0,
    max: 255
  };
  mutator = function(bytes) {
    var blue, green, red;
    red = bytes[0], green = bytes[1], blue = bytes[2];
    return {
      red: red,
      green: green,
      blue: blue
    };
  };
  return new Stochator(byte, byte, byte, mutator).next;
};

randomCharacter = function(lowercase) {
  var max, min, mutator, _ref;
  _ref = lowercase ? [97, 122] : [65, 90], min = _ref[0], max = _ref[1];
  mutator = function(charCode) {
    return String.fromCharCode(charCode);
  };
  return new Stochator({
    kind: "integer",
    min: min,
    max: max
  }, mutator).next;
};

/*

var chores = new Stochator({
    kind: "set",
    values: ["floors", "windows", "dishes"],
    replacement: false
});
var myChore = chores.next(); // "windows"
var yourChore = chores.next(); // "floors"
var hisChore = chores.next(); // "dishes"
var noOnesChore = chores.next(); // undefined


*/

lib/stochasm.js

;(function(globals){

//UMD
if (typeof define !== 'undefined' && define.amd) { //require.js / AMD
  define([], function() {
    return stochasm
  })
} else if (typeof module !== 'undefined' && module.exports) { //CommonJS
  module.exports = stochasm
} else { //script / browser
  globals.stochasm = stochasm
}


function stochasm() {
  var configs = [].slice.call(arguments)

  var generatorConfigs = configs.filter(function(c) {
    return typeof c === 'object'
  })

  var generator = {value: null} //value = initial value, rarely necessary
  generator.rand = Math.random //bring your own random function to the party if you want

  //mutator always last param
  var mutator = configs.slice(generatorConfigs.length)[0]
  generator.next = createGenerators.call(generator, generatorConfigs, mutator)
  return generator
}

function callFunctions(fns) {
  return fns.map(function(fn) { return fn() })
}

function randomBoundedFloat (min, max) {
  if (min == null) min = 0
  if (max == null) max = 1
  var spread = max - min
  return this.rand() * spread + min
}

function randomBoundedInteger (min, max) {
  if (min == null) min = 0
  if (max == null) max = 1
  var spread = 1 + max - min
  return Math.floor(this.rand() * spread) + min
}

function randomNormallyDistributedFloat (mean, stdev, min, max) {
  var seed = randomBoundedFloat.call(this)
  var float = inverseNormalCumulativeDistribution(seed) * stdev + mean
    
  if ((min != null) && (max != null))
    return Math.min(max, Math.max(min, float))
  else 
    return float
}

function randomSetMember(vals) {
  var that = this
  var max = vals.length - 1
  return vals[randomBoundedInteger.call(that, 0, max)]
}

function randomSetMemberWithoutReplacement (vals) {
  var that = this
  if (vals[0] == null) return void 0
  return vals.splice(randomBoundedInteger.call(that, 0, vals.length),1)
}

function randomWeightedSetMember (vals, weights) {
  var member = undefined
  var weightSum = 0
  var float = randomBoundedFloat.call(this)

  vals.forEach(function(value, index) {
    if (member) return
    var weight = weights[index]
    if (float <= weightSum + weight && float >= weightSum)
      member = value
      
    return weightSum += weight
  })
  return member
}

//http://home.online.no/~pjacklam/notes/invnorm/
function inverseNormalCumulativeDistribution(probability) {
  var high = probability > 0.97575, low = probability < 0.02425

  if (low || high) {
    var numCoefficients = [-7.784894002430293e-3, -3.223964580411365e-1, -2.400758277161838, -2.549732539343734, 4.374664141464968]
    var denomCoeffcients = [7.784695709041462e-3, 3.224671290700398e-1, 2.445134137142996, 3.754408661907416]
    var numMaxExponent = 5, denomMaxExponent = 4
    var coefficient = low ? 1 : -1
    var base = Math.sqrt(-2 * Math.log(low ? probability : 1 - probability))
  } else {
    var numCoefficients = [-3.969683028665376e1, 2.209460984245205e2, -2.759285104469687e2, 1.383577518672690e2, -3.066479806614716e1, 2.506628277459239]
    var denomCoeffcients = [-5.447609879822406e1, 1.615858368580409e2, -1.556989798598866e2, 6.680131188771972e1, -1.328068155288572e1]
    var numMaxExponent = 5, denomMaxExponent = 5
    var coefficient = probability - 0.5
    var base = Math.pow(coefficient, 2)
  }
    
  function mapMaxExp(maxExp) {
    return function(value, index) {
      return value * Math.pow(base, maxExp - index);
    }
  }
    
  var numerator = numCoefficients.map(mapMaxExp(numMaxExponent)).reduce(function(a,b) { return a + b}) //compute sum
  var denominator = denomCoeffcients.map(mapMaxExp(denomMaxExponent)).reduce(function(a,b) { return a + b}) + 1
  return coefficient * numerator / denominator;
}

function floatGenerator(min, max, mean, stdev) {
  var that = this
  if (mean && stdev) 
    return function() { return randomNormallyDistributedFloat.call(that, mean, stdev, min, max) }
  else 
    return function() { return randomBoundedFloat.call(that, min, max) } 
}

function integerGenerator (min, max) {
  var that = this
  if (min == null) min = 0
  if (max == null) max = 1    
  return function() { return randomBoundedInteger.call(that, min, max) }
}

function setGenerator(values, replacement, weights) {
  var that = this

  if (replacement == null) replacement = true
  if (weights == null) weights = null
  if (!values || !values.length) throw Error("Must provide a 'values' array for a set generator.")
    
  if (replacement) {
    if (weights) 
      return function() { return randomWeightedSetMember.call(that, values, weights) }
    else 
      return function() { return randomSetMember.call(that, values) }
  } else {
    return function() { return randomSetMemberWithoutReplacement.call(that, values) }
  }
}

function createGenerator(cfg) {
  cfg.kind = cfg.kind || "float"
  switch (cfg.kind) {
    case "float":
      return floatGenerator.call(this, cfg.min, cfg.max, cfg.mean, cfg.stdev)
    case "integer":
      return integerGenerator.call(this, cfg.min, cfg.max)
    case "set":
      return setGenerator.call(this, cfg.values, cfg.replacement, cfg.weights)
  }
}

function createGenerators (configs, mutator) {
  var _this = this
  configs[0] = configs[0] || {}
  var generators = configs.map(function(cfg) { return createGenerator.call(_this, cfg) })

  if (!mutator) {
    if (generators.length === 1)
      var callGenerators = function() { return callFunctions(generators)[0] }
    else
      var callGenerators = function() { return callFunctions(generators) }
  } else {
    if (generators.length === 1)
      var caller = function() { return callFunctions(generators)[0] }
    else
      var caller = function() { return callFunctions(generators) }

    var callGenerators = function() {
      return _this.value = mutator.call(_this, caller(), _this.value);
    }
  }

  return function(times) { //next(5)
    if (!times) return callGenerators()

    var results = []
    for (var i = 1; i <= times; ++i)
      results.push(callGenerators())
    return results
  }
}


})(this);









lib/stochator.js
// Generated by CoffeeScript 1.4.0
(function() {
  var Stochator, callFunctions, floatGenerator, integerGenerator, inverseNormalCumulativeDistribution, isType, randomBoundedFloat, randomBoundedInteger, randomCharacter, randomColor, randomNormallyDistributedFloat, randomSetMember, randomSetMemberWithoutReplacement, randomWeightedSetMember, setGenerator, shuffleSet,
    __slice = [].slice;

  isType = function(type) {
    return function(arg) {
      return Object.prototype.toString.call(arg) === ("[object " + type + "]");
    };
  };

  callFunctions = function(fns) {
    var fn, _i, _len, _results;
    _results = [];
    for (_i = 0, _len = fns.length; _i < _len; _i++) {
      fn = fns[_i];
      _results.push(fn());
    }
    return _results;
  };

  randomBoundedFloat = function(min, max) {
    var spread;
    if (min == null) {
      min = 0;
    }
    if (max == null) {
      max = 1;
    }
    spread = max - min;
    return Math.random() * spread + min;
  };

  randomBoundedInteger = function(min, max) {
    var spread;
    if (min == null) {
      min = 0;
    }
    if (max == null) {
      max = 1;
    }
    spread = 1 + max - min;
    return Math.floor(Math.random() * spread) + min;
  };

  randomColor = function() {
    var byte, mutator;
    byte = {
      kind: "integer",
      min: 0,
      max: 255
    };
    mutator = function(bytes) {
      var blue, green, red;
      red = bytes[0], green = bytes[1], blue = bytes[2];
      return {
        red: red,
        green: green,
        blue: blue
      };
    };
    return new Stochator(byte, byte, byte, mutator).next;
  };

  randomNormallyDistributedFloat = function(mean, stdev, min, max) {
    var float, seed;
    seed = randomBoundedFloat();
    float = inverseNormalCumulativeDistribution(seed) * stdev + mean;
    if ((min != null) && (max != null)) {
      return Math.min(max, Math.max(min, float));
    } else {
      return float;
    }
  };

  randomCharacter = function(lowercase) {
    var max, min, mutator, _ref;
    _ref = lowercase ? [97, 122] : [65, 90], min = _ref[0], max = _ref[1];
    mutator = function(charCode) {
      return String.fromCharCode(charCode);
    };
    return new Stochator({
      kind: "integer",
      min: min,
      max: max
    }, mutator).next;
  };

  randomSetMember = function(set) {
    var max;
    max = set.length - 1;
    return set.get(randomBoundedInteger(0, max));
  };

  randomSetMemberWithoutReplacement = function(set) {
    if (!set.get(0)) {
      return void 0;
    }
    set.length -= 1;
    return set.pop(randomBoundedInteger(0, set.length));
  };

  randomWeightedSetMember = function(set, weights) {
    var float, member, weightSum, _ref;
    _ref = [void 0, 0, randomBoundedFloat()], member = _ref[0], weightSum = _ref[1], float = _ref[2];
    set.each(function(value, index) {
      var weight;
      if (member) {
        return;
      }
      weight = weights.get(index);
      if (float <= weightSum + weight && float >= weightSum) {
        member = value;
      }
      return weightSum += weight;
    });
    return member;
  };

  inverseNormalCumulativeDistribution = function(probability) {
    var base, coefficient, denomCoeffcients, denomMaxExponent, denominator, high, low, mapMaxExp, numCoefficients, numMaxExponent, numerator, _ref, _ref1;
    high = probability > 0.97575;
    low = probability < 0.02425;
    if (low || high) {
      numCoefficients = new Set([-7.784894002430293e-3, -3.223964580411365e-1, -2.400758277161838, -2.549732539343734, 4.374664141464968]);
      denomCoeffcients = new Set([7.784695709041462e-3, 3.224671290700398e-1, 2.445134137142996, 3.754408661907416]);
      _ref = [5, 4], numMaxExponent = _ref[0], denomMaxExponent = _ref[1];
      coefficient = low ? 1 : -1;
      base = Math.sqrt(-2 * Math.log(low ? probability : 1 - probability));
    } else {
      numCoefficients = new Set([-3.969683028665376e1, 2.209460984245205e2, -2.759285104469687e2, 1.383577518672690e2, -3.066479806614716e1, 2.506628277459239]);
      denomCoeffcients = new Set([-5.447609879822406e1, 1.615858368580409e2, -1.556989798598866e2, 6.680131188771972e1, -1.328068155288572e1]);
      _ref1 = [5, 5], numMaxExponent = _ref1[0], denomMaxExponent = _ref1[1];
      coefficient = probability - 0.5;
      base = Math.pow(coefficient, 2);
    }
    mapMaxExp = function(maxExp) {
      return function(value, index) {
        return value * Math.pow(base, maxExp - index);
      };
    };
    numerator = numCoefficients.map(mapMaxExp(numMaxExponent)).sum();
    denominator = denomCoeffcients.map(mapMaxExp(denomMaxExponent)).sum() + 1;
    return coefficient * numerator / denominator;
  };

  shuffleSet = function(set) {
    var index, randomIndex, tmp, values, _i, _ref;
    values = set.copy();
    for (index = _i = _ref = values.length - 1; _ref <= 0 ? _i < 0 : _i > 0; index = _ref <= 0 ? ++_i : --_i) {
      randomIndex = randomBoundedInteger(0, index);
      tmp = values[index];
      values[index] = values[randomIndex];
      values[randomIndex] = tmp;
    }
    return values;
  };

  floatGenerator = function(min, max, mean, stdev) {
    if (mean && stdev) {
      return function() {
        return randomNormallyDistributedFloat(mean, stdev, min, max);
      };
    } else {
      return function() {
        return randomBoundedFloat(min, max);
      };
    }
  };

  integerGenerator = function(min, max) {
    if (min == null) {
      min = 0;
    }
    if (max == null) {
      max = 1;
    }
    return function() {
      return randomBoundedInteger(min, max);
    };
  };

  setGenerator = function(values, replacement, shuffle, weights) {
    var set, weightsSet;
    if (replacement == null) {
      replacement = true;
    }
    if (shuffle == null) {
      shuffle = false;
    }
    if (weights == null) {
      weights = null;
    }
    if (!values || !values.length) {
      throw Error("Must provide a 'values' array for a set generator.");
    }
    set = new Set(values);
    if (shuffle) {
      return function() {
        return shuffleSet(set);
      };
    } else if (replacement) {
      if (weights) {
        weightsSet = new Set(weights);
        return function() {
          return randomWeightedSetMember(set, weightsSet);
        };
      } else {
        return function() {
          return randomSetMember(set);
        };
      }
    } else {
      return function() {
        return randomSetMemberWithoutReplacement(set);
      };
    }
  };

  Stochator = (function() {
    var VERSION;

    VERSION = "0.3.1";

    function Stochator() {
      var configs;
      configs = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      this.setGenerator(configs);
    }

    Stochator.prototype.createGenerator = function(config) {
      var generator, max, mean, min, replacement, shuffle, stdev, values, weights, _ref;
      if ((_ref = config.kind) == null) {
        config.kind = "float";
      }
      generator = (function() {
        switch (config.kind) {
          case "float":
            min = config.min, max = config.max, mean = config.mean, stdev = config.stdev;
            return floatGenerator(min, max, mean, stdev);
          case "integer":
            return integerGenerator(config.min, config.max);
          case "set":
            values = config.values, replacement = config.replacement, shuffle = config.shuffle, weights = config.weights;
            return setGenerator(values, replacement, shuffle, weights);
          case "color":
          case "rgb":
            return randomColor(config.kind);
          case "a-z":
          case "A-Z":
            return randomCharacter(config.kind === "a-z");
        }
      })();
      if (!generator) {
        throw Error("" + config.kind + " not a recognized kind.");
      } else {
        return generator;
      }
    };

    Stochator.prototype.createGenerators = function(configs, mutator) {
      var callGenerators, caller, config, generators, _ref,
        _this = this;
      if ((_ref = configs[0]) == null) {
        configs[0] = {};
      }
      generators = (function() {
        var _i, _len, _results;
        _results = [];
        for (_i = 0, _len = configs.length; _i < _len; _i++) {
          config = configs[_i];
          _results.push(this.createGenerator(config));
        }
        return _results;
      }).call(this);
      if (!mutator) {
        callGenerators = generators.length === 1 ? function() {
          return callFunctions(generators)[0];
        } : function() {
          return callFunctions(generators);
        };
      } else {
        caller = generators.length === 1 ? function() {
          return callFunctions(generators)[0];
        } : function() {
          return callFunctions(generators);
        };
        callGenerators = function() {
          return _this.value = mutator.call(_this, caller(), _this.value);
        };
      }
      return function(times) {
        var time, _i, _results;
        if (times) {
          _results = [];
          for (time = _i = 1; 1 <= times ? _i <= times : _i >= times; time = 1 <= times ? ++_i : --_i) {
            _results.push(callGenerators());
          }
          return _results;
        } else {
          return callGenerators();
        }
      };
    };

    Stochator.prototype.setGenerator = function(configs) {
      var config, generatorConfigs, mutator, name, _i, _len, _ref, _ref1;
      generatorConfigs = [];
      for (_i = 0, _len = configs.length; _i < _len; _i++) {
        config = configs[_i];
        if (isType("Object")(config)) {
          generatorConfigs.push(config);
        } else {
          break;
        }
      }
      _ref = configs.slice(generatorConfigs.length), name = _ref[0], mutator = _ref[1];
      name || (name = "next");
      if (isType("Function")(name)) {
        _ref1 = ["next", name], name = _ref1[0], mutator = _ref1[1];
      }
      return this[name] = this.createGenerators(generatorConfigs, mutator);
    };

    Stochator.prototype.toString = function() {
      return "[object Stochator]";
    };

    return Stochator;

  })();

  if (typeof module !== "undefined" && module !== null ? module.exports : void 0) {
    module.exports = Stochator;
  } else {
    this.Stochator = Stochator;
  }

}).call(this);
